### PR TITLE
Use less calls for MoveDirectory

### PR DESCRIPTION
**Pull request by m0ddixx**
reopened because I was not paying attention to workflow... :man_facepalming:

See #58240
This fix is trying to overcome the concept of matching a universal behaviour inside a platform aware implementation.

All checks for consistent behaviours can be done within the public Directory implementation.
The missing bits which are only relevant for Unix are being left e.g. matching trailing seperators on both paths.

This might be redundant, given that Windows would handle such cases, so we dont have to do the checks beforehand, but given that we can reduce calls on both Unix and Windows runtimes should be a slight improvement.
There is only a single FileExist call for Unix and for windows we won't execute the move operation and wait for exceptions.

### CODE REVIEW

**Comment by m0ddixx** [comment]
tests on wasm probably failing due to #40536
I don't see any reason, why no IOException is being thrown when source path is file and ends with separator

File: src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs [hunk 1]
```diff
@@ -271,6 +271,7 @@ public static void Move(string sourceDirName, string destDirName)
             bool sameDirectoryDifferentCase =
                 directoriesAreCaseVariants &&
                 destDirNameFromFullPath.Equals(sourceDirNameFromFullPath, fileSystemSensitivity);
+            bool sourcePathIsFile = FileSystem.FileExists(fullsourceDirName);
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs:274]: Seems like you want to hit the file system once to get both whether it's an existing directory, and whether it's an existing file.

@stephentoub I can't remember how one does this without pinvoke. If I create a DirectoryInfo and it's a file, it says exists is false. However, if the access time is not default, it is a file that exists. With FileInfo, I can do the same in reverse. Strange behavior and surely there's a better way..?

File: src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs [hunk 2]
```diff
@@ -285,9 +286,15 @@ public static void Move(string sourceDirName, string destDirName)
...
             // to make sure our cross platform behavior matches .NET Framework behavior.
-            if (!FileSystem.DirectoryExists(fullsourceDirName) && !FileSystem.FileExists(fullsourceDirName))
+            if (!FileSystem.DirectoryExists(fullsourceDirName) && !sourcePathIsFile)
                 throw new DirectoryNotFoundException(SR.Format(SR.IO_PathNotFound_Path, fullsourceDirName));
 
+            // To match API specification, we check if the source path is a file and ends with a trailing slash.
+            // While having a file as source is no problem for any OS, Windows will throw an error if the source path has a trailing slash.
+            // To match platform agnostic behaviour, we check that before handing over to OS specific implementation.
+            if (sourcePathIsFile && Path.EndsInDirectorySeparator(fullsourceDirName))
+                throw new IOException(SR.Format(SR.IO_PathNotFound_Path, fullsourceDirName));
```
Comment (carlossanlop) [inline src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs:296]: Is this case unit testable?

File: src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs (hunk 1 above)
Comment (carlossanlop) [inline src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs:274]: Why call it this early? Consider moving this new call right before the `if` where it is first used (line 289).

File: src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs (hunk 2 above)
Comment (carlossanlop) [inline src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs:296]: Is this case unit testable?

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs [hunk 3]
```diff
@@ -372,23 +372,10 @@ public static void CreateDirectory(string fullPath)
...
-
-                if (Path.EndsInDirectorySeparator(sourceFullPath))
-                    throw new IOException(SR.Format(SR.IO_PathNotFound_Path, sourceFullPath));
-
-                // ... but it doesn't care if the destination has a trailing separator.
+            // We have to check, if the source path does not end with a trailing separator.
+            // To match windows behaviour, Unix does not want trailing separators for both paths when it moves a file.
+            if (!Path.EndsInDirectorySeparator(sourceFullPath))
                 destFullPath = Path.TrimEndingDirectorySeparator(destFullPath);
-            }
```
Comment (carlossanlop) [inline src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:391]: I'm not convinced about changing this behavior. This code was originally added in PR https://github.com/dotnet/corefx/pull/19716 to fix this issue https://github.com/dotnet/runtime/issues/21673 .

**Review by carlossanlop — COMMENTED** [review]
Thanks for your PR, @m0ddixx . I left some comments/questions for you to consider.

File: src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs (hunk 2 above)
Comment (m0ddixx) [reply src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs:296]: Should already be present
https://github.com/dotnet/runtime/blob/206dba75f00e67c839709e0632b22819329d73d0/src/libraries/System.IO.FileSystem/tests/Directory/Move.cs#L96-L102

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs (hunk 3 above)
Comment (m0ddixx) [reply src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:391]: If the code is untouched, we end up hitting FileExists twice on Unix. The checks will happen before and once we get to FileSystem.Unix.Move we saved a few round trips. But the behaviour should be the same as before

**Comment by jeffhandley** [comment]
@dotnet/area-system-io This could use another review; the remaining concern/feedback doesn't look actionable as it is.

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs (hunk 3 above)
Comment (carlossanlop) [reply src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:391]: Also, the change does not make sense: The new code is checking if _source path_ ends in separator, but then trimming the _destination path_.

File: src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs [hunk 4]
```diff
@@ -283,11 +283,19 @@ public static void Move(string sourceDirName, string destDirName)
             if (!sourceRoot.Equals(destinationRoot, StringComparison.OrdinalIgnoreCase))
                 throw new IOException(SR.IO_SourceDestMustHaveSameRoot);
 
+            bool sourcePathIsFile = FileSystem.FileExists(fullsourceDirName);
```
Comment (carlossanlop) [inline src/libraries/System.Private.CoreLib/src/System/IO/Directory.cs:286]: The problem of caching if the file exists, is that the file could exist initially, but then between this call and the subsequent places where the cached value is checked, the file could end up being deleted.

**Comment by carlossanlop** [comment]
@m0ddixx thank you so much for your contribution, and apologies for the delay.

I don't think the proposed change is effectively reducing the calls without changing the behavior, for the following reasons:

- Caching if a file exists is not a good idea since the existence of the file could exist between the caching action and checking the cached value.

- The proposed change is checking for a separator in one path, then removing the separator in a different path.

- The original issue does point out there are multiple `FileExists` calls, but notice that the passed path is the source path in the first, and the destination path in the second. So they aren't necessarily related.

I think there's no action for us to take here, mainly due to the third point, so I'll go ahead and close both the issue and this PR.

**Pull request closed without merging**